File: Algorithms/Complete Search/Generating Subsets/Algorithm1.cpp
#include <bits/stdc++.h>
#include <time.h>
using namespace std;
//Method 1- Recursive Search returns all subsets of {0,1,...n-1}
int n;
vector<int> subsets;
void search(int k)
{
  if(k==n)
  {
    if(!subsets.size())
     cout<<"empty\n";
    else
    {
      cout<<'[';
      for(auto x:subsets)
        cout<<x <<',';
      cout<<"\b]\n";
    }
  }
  else
  {
    search(k+1); //this search doesnt get k
    subsets.push_back(k);
    search(k+1); //this search gets k
    subsets.pop_back();
  }
}
int main()
{
  srand(time(0));
  ::n=rand()%5;
  cout<<"n=" <<::n <<endl;
  search(0);
}
//very beautiful O(2^n) algorithm

File: Algorithms/Complete Search/Generating Subsets/Algorithm1.py
#Method 1- Recursive Search returns all subsets of {0,1,...n-1}
import random
n=random.randint(0,5)
print(n)
subset=[]
def search(k):
    if k==n:
        print(subset)
    else:
        search(k+1)
        subset.append(k)
        search(k+1)
        subset.pop()
search(0)
